src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryResult.cs

```diff
@@ -15,29 +16,62 @@ public sealed class PluginDiscoveryResult
         /// </summary>
         public PluginFile PluginFile { get; }
 
+        private string _message;
+
         /// <summary>
         /// Gets a message if <see cref="PluginFile.State" /> is not <see cref="PluginFileState.Valid" />;
         /// otherwise, <c>null</c>.
         /// </summary>
-        public string Message { get; }
+        public string Message
+        {
+            get
+            {
+                if (_message == null)
+                {
+                    switch (PluginFile.State.Value)
+                    {
+                        case PluginFileState.Valid:
+                            break;
+
+                        case PluginFileState.NotFound:
+                            _message = string.Format(
+                                CultureInfo.CurrentCulture,
+                                Strings.Plugin_FileNotFound,
+                                PluginFile.Path);
+                            break;
+
+                        case PluginFileState.InvalidFilePath:
+                            _message = string.Format(
+                                CultureInfo.CurrentCulture,
+                                Strings.Plugin_InvalidPluginFilePath,
+                                PluginFile.Path);
+                            break;
+
+                        case PluginFileState.InvalidEmbeddedSignature:
+                            _message = string.Format(
+                                CultureInfo.CurrentCulture,
+                                Strings.Plugin_InvalidEmbeddedSignature,
+                                PluginFile.Path);
+                            break;
+
+                        default:
+                            throw new NotImplementedException();
```

Comment:
NotImplementedException doesn't sound like the right choice here, unless I am misunderstanding. InvalidOperationException instead maybe?

Reply:
Hmm...I get where you going but from an API perspective, it's just not implemented. 
It's up to the author(us) to make sure all cases are handled. 

InvalidOperationException to me suggests that there might be something the consumer (whether API or NuGet client) can do to fix the problem.